pkg/boot/universalpayload/fitparser.go

```diff
@@ -0,0 +1,268 @@
+package universalpayload
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+)
+
+const (
+	FitBeginNode = 0x1
+	FitEndNode   = 0x2
+	FitDrop       = 0x3
+	FitNop        = 0x4
+	FitEnd        = 0x9
+)
+
+// FitHeader represents the header of an FIT file.
+type FitHeader struct {
+	Magic           uint32
+	TotalSize       uint32
+	OfFITStruct     uint32
+	OfFITStrings    uint32
+	OffMemRsvmap    uint32
+	Version         uint32
+	LastCompVersion uint32
+	BootCpuidPhys   uint32
+	SizeDtStrings   uint32
+	SizeDtStruct    uint32
+}
+
+type Property struct {
+	Name  string
+	Value []byte
+}
+
+type Node struct {
+	Name       string
+	Properties []Property
+	Children   []*Node
+}
+
+func ReadFitHeader(data []byte) (*FitHeader, error) {
+	if len(data) < 40 {
+		return nil, errors.New("data too short to be a valid FIT header")
+	}
+
+	header := &FitHeader{
+		Magic:           binary.BigEndian.Uint32(data[0:4]),
+		TotalSize:       binary.BigEndian.Uint32(data[4:8]),
+		OfFITStruct:     binary.BigEndian.Uint32(data[8:12]),
+		OfFITStrings:    binary.BigEndian.Uint32(data[12:16]),
+		OffMemRsvmap:    binary.BigEndian.Uint32(data[16:20]),
+		Version:         binary.BigEndian.Uint32(data[20:24]),
+		LastCompVersion: binary.BigEndian.Uint32(data[24:28]),
+		BootCpuidPhys:   binary.BigEndian.Uint32(data[28:32]),
+		SizeDtStrings:   binary.BigEndian.Uint32(data[32:36]),
+		SizeDtStruct:    binary.BigEndian.Uint32(data[36:40]),
+	}
+
+	fmt.Printf("Magic: 0x%x\n", header.Magic)
+	if header.Magic != 0xd00dfeed {
+		return nil, errors.New("invalid FIT magic number")
+	}
+
+	return header, nil
+}
```

Comment:
I think this can be simpler: https://go.dev/play/p/V52qzPxrjeg

func ReadFitHeader(r io.Reader) (*FitHeader, error) {
	var header FitHeader
	if err := binary.Read(r, binary.LittleEndian, &header); err != nil {
		return nil, err
	}

	if header.Magic != 0xd00dfeed {
		return nil, errors.New("invalid FIT magic number")
	}

	return &header, nil
}

in any event, I think an io.Reader is preferable to a []byte, since they can be a bit fragile.